src/Component/Router.purs

```diff
@@ -43,29 +42,15 @@ data Query a
 type Input =
   Maybe Route
 
--- If you haven't seen nested `Coproduct` or `Either` before, or you haven't worked with multiple types of
--- child component, then these types are probably confusing. They're a little tedious to define and are
--- being removed in favor of a much nicer mechanism in Halogen 5, but are necessary in Halogen 4.
--- 
--- For a detailed explanation of what's going on here, please see this issue:
--- https://github.com/thomashoneyman/purescript-halogen-realworld/issues/20
-type ChildQuery = Coproduct7
-  Home.Query
-  Login.Query
-  Register.Query
-  Settings.Query
-  Editor.Query
-  ViewArticle.Query
-  Profile.Query
-
-type ChildSlot = Either7
-  Unit
-  Unit
-  Unit
-  Unit
-  Unit
-  Unit
-  Unit
+type ChildSlots = 
```

Comment:
In Halogen 5 we use a row type to describe child components, instead of the old nested either / coproduct pairing. It's quite a bit cleaner and allows you to easily give descriptive names to each component.

A child path used to be a pairing of a component's query type (what a parent can use to trigger behaviors in or request information from one of its child components) and slot type (a unique address when there are multiple components of the same type). 

In Halogen 5, we instead have a dedicated `Slot` type which uses the child component's query type, message type (what outputs are possible to handle from the component), and slot address.

For example:

```purescript
-- this parent component has one child component type, labeled `component`, with the query
-- type `ComponentQuery`, outputs of type `ComponentMessage`, and we'll distinguish among
-- multiple copies of the component using an `Int` address.
type MyChildSlots =
  ( component :: H.Slot ComponentQuery ComponentMessage Int )
```

Our router deals with page components which raise no output messages and have no public query interface you can use to request information or manipulate them. They simply aren't necessary. So for each of our page components the slot type is just:

```purescript
type ChildSlots =
  ( home :: H.Slot (Const Void) Void Unit )
```

In other words, `H.Slot` expects a query type of kind `Type -> Type`, a message type of kind `Type`, and a slot address of kind `Type`. When you need to provide a type but have no possible values that could satisfy the type, you can use `Void` -- a type with no accompanying values. And when you need to provide something of kind `Type -> Type` but have no possible values, you can use `Const Void`. So when we have a component with no queries and no messages we can use `Const Void` and `Void` to represent that.

This is fairly common in our application, so I've defined a helper type in `Component.Utils` for components with no queries or messages:

```purescript
type OpaqueSlot = H.Slot (Const Void) Void
```